### DISCUSSION TITLE
ly2: space savings

### DISCUSSION THREAD

**Discussion by choppsv1** [CESNET/libyang#1545] [Ideas]
In some other discussions the issue of space usage came up. There's a couple very quick fixes for saving some space that could be made safely prior to API freeze.

1) group integers together and not between pointers.
2) pack enumerations.

For 1) there's a lot of:
```
    ...
    const char *value;              /**< original value */
    LY_VALUE_FORMAT format;        /**< format of the node and any prefixes, ::LY_VALUE_XML or ::LY_VALUE_JSON */
    void *val_prefix_data;          /**< format-specific prefix data */
    uint32_t hints;                 /**< additional information about from the data source, see the [hints list](@ref lydhints) */
    struct lyd_attr *attr;          /**< pointer to the list of generic attributes of this node */
    ...
```
Most people are running foo/arm64 which will be a 32/64 ABI so both the above is:
```
0: value
8: format
16: val_prefix_data
24: hints
32: attr
40:
```
rearrange this to group the ints at the bottom and you get:
```
0: value
8: val_prefix_data
16: attr
24: format
28: hints
32:
```

pack the enumerations, change hints to uint16_t (which seems by the defines to be it's target size) and arrange again and you get:

```
0: value
8: val_prefix_data
16: attr
24: hints
26: format
27:
```

**Comment by choppsv1** [comment]
FWIW, the grouping is more important than the bottom part. Especially if you are going past 64 bytes to get to them. Almost everyone has 64B caches now, so everything in the first 64B has the same access time basically. This is also a good reason to group, to avoid going past the 64B size for anything at all if possible.

**Comment by michalvasko** [comment]
Right, now is probably the time to reorder struct members, it was useless when they were still changing.
